Adamant/Modules/Chat/View/ChatViewController.swift

```diff
@@ -189,11 +194,29 @@ final class ChatViewController: MessagesViewController {
         super.collectionView(collectionView, willDisplay: cell, forItemAt: indexPath)
     }
     
+    override func scrollViewDidEndDragging(_: UIScrollView, willDecelerate _: Bool) {
+        viewModel.didEndScroll()
+    }
+    
     override func scrollViewDidScroll(_ scrollView: UIScrollView) {
         super.scrollViewDidScroll(scrollView)
         updateIsScrollPositionNearlyTheBottom()
         updateScrollDownButtonVisibility()
         
+        let targetY: CGFloat = 20 + view.safeAreaInsets.top
+        guard let visibleIndexPaths = messagesCollectionView.indexPathsForVisibleItems.sorted(by: {
+            $0.row < $1.row
+        }) as [IndexPath]? else { return }
+        for indexPath in visibleIndexPaths {
+            if let cell = messagesCollectionView.cellForItem(at: indexPath) {
+                let cellRect = messagesCollectionView.convert(cell.frame, to: self.view)
+                if cellRect.minY <= targetY && cellRect.maxY >= targetY {
+                    viewModel.checkTopMessage(indexPath: indexPath)
+                    break
+                }
+            }
+        }
```

Comment:
Move it to a separate method like `updateScrollDownButtonVisibility()`

Reply:
use `guard` to improve code readability